tv_apps/smart-home/js/edit.js

```diff
@@ -104,57 +140,109 @@
       },
       getLength: function() {
         return this.queue.length;
+      },
+      clearQueue: function() {
+        this.queue.length = 0;
       }
     },
 
-    _swapTimer: undefined,
+    _moveTimer: undefined,
 
-    _produceSwap: function(key) {
-      this._swapQueue.push(key);
-      if (!this._swapTimer) {
-        this._consumeSwap();
+    _produceMove: function(key) {
+      this._moveQueue.push(key);
+      if (!this._moveTimer) {
+        this._consumeMove();
       }
     },
 
-    _consumeSwap: function() {
+    _consumeMove: function() {
       var focus = this.spatialNavigator.getFocusedElement();
       var keepConsuming = false;
-      if (!this._swapQueue.isEmpty() && focus.CLASS_NAME === 'XScrollable') {
-        var key = this._swapQueue.shift();
+
+      if (!this._moveQueue.isEmpty() && focus.CLASS_NAME === 'XScrollable') {
+        var key = this._moveQueue.shift();
         var targetItem = focus.getTargetItem(key);
-        if (targetItem) {
-          var focusedCard = this.cardManager.findCardFromCardList(
-                                  {cardId: focus.currentItem.dataset.cardId});
-          var nonFocusedCard = this.cardManager.findCardFromCardList(
-                                    {cardId: targetItem.dataset.cardId});
-          this.cardManager.swapCard(focusedCard, nonFocusedCard);
-          this._swapTimer =
-            window.setTimeout(this.onCardSwapAnimationEnd.bind(this),
-              CARD_TRANSFORM_LATENCY, targetItem);
-        } else {
-          // if queue is not empty but we are failed to get targetItem
-          // that means we reach the boundary of cards. Thus we should
-          // consume next action in queue
-          keepConsuming = true;
+
+        switch (key) {
+        case 'left':
+        case 'right':
+          if (!this._isHovering && targetItem &&
+              focus.currentItem.getAttribute('app-type') !== 'folder' &&
+              targetItem.getAttribute('app-type') === 'folder') {
+            // hover on folder
+            this.hoverCard(focus, focus.currentItem, targetItem);
+            this._moveTimer =
+              window.setTimeout(this.onCardMoveAnimationEnd.bind(this),
+                CARD_TRANSFORM_LATENCY, targetItem);
+          } else if (this._isHovering &&
+                     (this._previousKey === 'left' && key === 'right' ||
+                      this._previousKey === 'right' && key === 'left')) {
```

Comment:
How about remembering `hoverFrom` (`'left'` or `'right'`) in `scrollable` and remove `_previousKey` here?